File: test/scripts/23-thread-pools/incorrect.py
#!/usr/bin/env python3
"""
INCORRECT: Create new thread per task - overhead.
Resume: "Use Thread Pools"
"""
import os
import signal
import threading
import time
TASKS = 500
DURATION_SEC = int(os.environ.get("DURATION_SEC", "60"))
stop = False

def task():
    return sum(range(1000))

def main():
    global stop
    print("INCORRECT: new Thread per task")
    signal.signal(signal.SIGINT, lambda s, f: globals().__setitem__('stop', True))
    start = time.perf_counter()
    laps = 0
    while not stop and (time.perf_counter() - start) < DURATION_SEC:
        threads = [threading.Thread(target=task) for _ in range(TASKS)]
        for t in threads:
            t.start()
        for t in threads:
            t.join()
        laps += 1
    print(f"Stopped: {laps} laps")

if __name__ == "__main__":
    main()

File: test/scripts/23-thread-pools/incorrect.js
const { Worker } = require('worker_threads');
const path = require('path');
const TASKS = 100;
const DURATION_SEC = parseInt(process.env.DURATION_SEC || '60', 10);
const code = `let t=0;for(let i=0;i<1000;i++)t+=i;`;
let laps = 0;
const start = Date.now();
console.log('INCORRECT: new Worker per task');
function run() {
  if (Date.now() - start >= DURATION_SEC * 1000) {
    console.log('Stopped:', laps, 'laps');
    return;
  }
  const workers = [];
  for (let i = 0; i < TASKS; i++) {
    workers.push(new Promise((r) => {
      const w = new Worker(code, { eval: true });
      w.on('exit', r);
    }));
  }
  Promise.all(workers).then(() => { laps++; setImmediate(run); });
}
run();
process.on('SIGINT', () => process.exit(0));
